geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/command/CommitCommand.java

```diff
@@ -97,10 +112,35 @@ public void cmdExecute(final Message clientMessage, final ServerConnection serve
         commitMsg.setClientVersion(null); // fixes bug 46529
       }
     }
+    if (txException != null) {
+      DistributedMember target = txProxy.getTarget();
+      // a TransactionInDoubtException caused by the TX host shutting down means that
+      // the transaction may still be active and hold locks. We must wait for the transaction
+      // host to finish shutting down before responding to the client or it could encounter
+      // conflicts in retrying the transaction
+      if ((txException instanceof TransactionInDoubtException)
+          && (txException.getCause() instanceof CancelException)) {
+        logger.info(
+            "Waiting for departure of {} before throwing TransactionInDoubtException.",
+            target);
+        try {
+          serverConnection.getCache().getDistributionManager().getMembershipManager()
+              .waitForDeparture(target);
+        } catch (TimeoutException e) {
+          // status will be logged below
+        } catch (InterruptedException e) {
+          return;
```

Comment:
We still need to send the TransactionInDoubtException back to client in this case to avoid client hang waiting for the reply.

Reply:
If the thread is interrupted we ought to be shutting down, no?  I could add a cancellation check if that would set your mind at ease.

Reply:
If only shutdown could cause the thread to be interrupted, it is fine to return here.